### PR TITLE
[megatron] compute logprobs/entropy on CP sharded logits + fix CP loss scaling

### PR DESCRIPTION
**Pull request by jinghanyao1-hub**
# PR README: Memory-safe Megatron packed logprobs

## Problem

With Megatron context parallelism and packed microbatches enabled, the old wrapper path
postprocessed packed logits back into dense batch form before computing token logprobs:

```text
packed CP logits [1, T_local, V_tp]
  -> postprocess_packed_seqs(...)
  -> dense logits [B, S, V_tp]
  -> selected token logprobs [B, S-1]
```

That dense `[B, S, V_tp]` materialization causes a large transient memory spike.

## 8k memory comparison

At the same 8k setup on `Qwen/Qwen3-4B-Instruct-2507`, the new packed-logprob path removes the
dense-logit spike:

| path | profile | peak allocated | peak reserved |
| --- | --- | ---: | ---: |
| old dense packed-logit path | 8k CUDA memory snapshot | 41.31 GiB | 47.06 GiB |
| new packed-logprob path | 8k CUDA memory snapshots, max across ranks | 29.97 GiB | 30.03 GiB |

This is an 11.34 GiB allocated-memory reduction and a 17.03 GiB reserved-memory reduction at 8k.
Both numbers were reconstructed from PyTorch CUDA memory snapshots of the policy forward/backward
region.

## Test setup

Benchmarks used:

```text
model: Qwen/Qwen3-4B-Instruct-2507
node config: 1 node, 8 GPUs
Megatron parallelism: TP=1, PP=1, CP=8
policy/ref placement: 8 GPUs per node, colocate_all=true
CP communication: p2p and a2a
micro train/forward batch size per GPU: 1
train batch size: 1
response length: 1024
```

## Change

The new wrapper path keeps logits in packed/context-parallel form and computes only the selected
token logprobs:

```text
packed CP logits [1, T_local, V_tp]
  + packed target ids [1, T_padded]
  -> from_parallel_logits_to_logprobs_packed_sequences(...)
  -> packed scalar logprobs [B, S-1]
  -> action logprobs [B, num_actions]
```

Concretely, when `remove_microbatch_padding` creates `packed_seq_params`, the wrapper now:

1. Builds matching packed target token IDs.
2. Calls `from_parallel_logits_to_logprobs_packed_sequences(...)`.
3. All-gathers only scalar logprobs across CP ranks.
4. Skips `postprocess_packed_seqs(...)` for packed logits.

The normal non-packed logprob path is unchanged.

This packed scalar-logprob feature now applies to both Megatron RL/default training and Megatron
SFT sequence-packed training.

## Follow-up change: SFT packed path

The same packed scalar-logprob path is now used by SFT as well as RL. This keeps SFT sequence
packing active, but avoids unpacking hidden states or logits through `postprocess_packed_seqs(...)`.

For packed SFT microbatches, the worker now:

1. Computes selected-token logprobs/entropy while tensors are still packed.
2. Scatters packed scalar outputs back to per-sample positions.
3. Avoids materializing dense `[B, S, V]` logits or dense unpacked hidden states.

After this change, the old packed-output postprocessing helper was no longer on a live path and was
removed along with its obsolete round-trip tests.

## Validation

```text
tests/backends/skyrl_train/gpu/gpu_ci/megatron/test_megatron_worker.py
  test_megatron_forward[cp_2_policy_seq_packing]
  test_megatron_forward[cp_2_policy_seq_packing_a2a]
  test_megatron_forward[tp_2_cp_2_policy_seq_packing]
  test_megatron_train[tp2_cp2_policy_seq_packing_no_entropy_loss]
  test_megatron_train[tp2_cp2_policy_seq_packing_with_entropy_loss]
  test_megatron_train[tp2_cp2_policy_seq_packing_no_entropy_loss_a2a]
  result: 6 passed

tests/backends/skyrl_train/distributed/test_preprocess_packed_seqs_multiseq.py
tests/train/test_packing_round_trip.py
  result: 12 passed

tests/backends/skyrl_train/gpu/gpu_ci/megatron/test_sft_packing_parity.py
  result: 1 passed
  packed DP=2: PASS, avg diff 1.6029e-02
  packed CP=2/DP=1: PASS, avg diff 3.0107e-02
```

Also ran `examples/train/megatron/run_megatron.sh` on a 4-train/8-val
GSM8K subset with `max_generate_length=128`.

```text
CP=1: completed, Training done
  mesh: world_size=8, TP=2, PP=2, DP=2, CP=1
  eval/all/avg_score: 0.0000
  rollout/train logprob diff: max 0.2596, mean 0.0149

CP=2: completed, Training done
  mesh: world_size=8, TP=2, PP=2, DP=1, CP=2 via sp=0/1
  eval/all/avg_score: 0.0000
  rollout/train logprob diff: max 0.2596, mean 0.0143
```

## 128k result

After removing the 8k dense-logit memory spike, the patched path completed the final CP benchmark
sweep through 128k total length:

At 128k total tokens (`max_total_len=131072`, prompt `130048`, response `1024`):

```text
p2p peak allocated: 53.53 GiB
p2p peak reserved:  57.26 GiB
a2a peak allocated: 53.53 GiB
a2a peak reserved:  56.68 GiB
```

So the patch reduces 8k peak memory substantially and enables clean 128k packed Megatron runs,
without gathering dense `[B, S, V]` logits.

### CODE REVIEW

**Comment by jinghanyao1-hub** [comment]
Note, since `postprocess_packed_seqs` has been removed, some tests also need to be updated as done in this PR.

**Review by erictang000 — COMMENTED** [review]
this looks really good! thanks for all the improvements and adding the CP sharded entropy loss calculations, I think we are pretty close to being able to merge this pending a run of all megatron GPU CI.

The one thing that is outstanding is some E2E runs that are not just smoke tests

Image

it would be great to rerun the above experiments with just the full `run_megatron.sh` script and get validated reward curves. In addition to those two full runs (with and without CP on this PR), it would be good to do a run with a2a CP, and one without CP on the current main branch to compare that the new entropy calculations match with previous values.

So 4 e2e runs in total (Qwen3-0.6B GSM8k with `run_megatron.sh`):
1. no CP on this branch
2. with CP ring attn on this branch
3. with CP a2a on this branch
4. no CP on main (to check entropy)

**Comment by jinghanyao1-hub** [comment]
Running some e2e train to further verify the update...

**Comment by erictang000** [comment]
tracking gpu ci in this job: https://console.anyscale.com/jobs/prodjob_pmr4cmyqq3n8q73e3kbe8dexs6

**Comment by jinghanyao1-hub** [comment]
Verified on Qwen3-0.6B on GSM8K with GRPO recipe, training batch size 128, eval at iteration 20:

### Global Step 20

| Run | eval pass@1 / avg score | avg generated tokens |
| --- | ---: | ---: |
| PR no-CP | `0.682335` | `489.553` |
| main no-CP | `0.697498` | `493.403` |
| PR CP ring | `0.695982` | `487.799` |
| PR CP a2a | `0.681577` | `513.346` |

Average no-CP eval pass@1: `0.689917`

Average CP eval pass@1: `0.688779`

Delta CP - no-CP: `-0.001137`

**Comment by jinghanyao1-hub** [comment]
Also found a loss scale issue with the current main branch, causing the CP training always doubling the loss.
Screenshot 2026-06-04 at 8 35 53 PM

The CP grad-norm double-counts CP loss scaling. SkyRL’s loss in [megatron_model_wrapper.py](https://github.com/NovaSky-AI/SkyRL/blob/42374592ec228c5404578780b45a186cae52e41e/skyrl/backends/skyrl_train/workers/megatron/megatron_model_wrapper.py#L372) returns the two-output loss form, so Megatron handles it in `megatron/core/pipeline_parallel/schedules.py` by multiplying the returned loss by `cp_group_size` before dividing by `num_microbatches`. DDP then averages gradients over the DP+CP group. Therefore the Megatron CP schedule multiplier already cancels the CP part of DDP averaging, so SkyRL should use DP-only in `grad_sum_correction_factor` and should not apply an additional `* cp_size` to KL/entropy. After removing the extra CP scaling, CP-ring/a2a grad norms match the no-CP range.

**Comment by erictang000** [comment]
WandB report with runs: https://wandb.ai/sky-posttraining-uc-berkeley/skyrl_cp_packed_logprobs_e2e?nw=nwusereatang

**Review by erictang000 — APPROVED** [review]
looks good to me!

huge first PR to SkyRL 🚀🚀🚀🚀
